### ISSUE TITLE
Use canvas-like width & height

### ISSUE THREAD

**Issue by jakearchibald** [WICG/portals#173]
With canvas, the 'viewport' size is specified by the width/height attributes, whereas the size specified by CSS is the rendered size. Is that a more useful model for portals? Would it mean developers wouldn't need to mess around with scaling so much?

Eg:

```html
<portal src="…" width="1920" height="1080" style="width: 100%; height: auto">
```

This portal would display the src rendered at 1920x1080, but its layout on the page would be the width of the parent element, and the height would be `widthOfParent/(1920/1080)` (maintaining aspect ratio).

The rendering of the element would be scaled to fit, according to `object-fit` and `object-position`.

**Comment by jeremyroman** [comment]
Co-opting the issue, if you don't mind. :)

The question here is how portals should be resized (in the outer document layout, and the viewport size of the browsing context). The iframe model is that the layout size defines the viewport size, and updates to the layout size in turn update the viewport size.

Since portals expect to eventually be activated (at which point its viewport size should match the outer viewport), it seems appealing to fix the viewport size to match. This also prevents the host document from using its control over viewport size (visible to the portal contents) to communicate. What do people think of that?

If we do that, then we've diverged from the iframe model, and the natural choice is to use the replaced object model (presumably with the viewport size as the intrinsic size, and allowing the author to use `width`, `height`, `object-fit`, etc to control the layout size and the scaling between the two). In that world we could either eliminate the `width` and `height` properties or leave them as presentational attributes (which essentially control the CSS `width` and `height` properties, like for `<img>`).

**Comment by domenic** [comment]
I like that plan, although I think it makes #3 more important, as with that model it's exceedingly unlikely that a page would look reasonable out of the box when portaled. (Whereas before it might show, e.g., its responsive mobile layout.)

If implemented, I would remove `with` and `height`.

**Comment by jeremyroman** [comment]
I don't follow your reasoning. Why wouldn't it look reasonable when portaled (and, presumably, scaled by 0.1x or something)?

**Comment by domenic** [comment]
I guess both:

<img src="https://user-images.githubusercontent.com/617481/81337159-96e09f80-9078-11ea-9d0d-8433dd5eae18.png" width="180">

and:

<img src="https://user-images.githubusercontent.com/617481/81337302-cabbc500-9078-11ea-9aac-b2f633dec42e.png" width="180">

are pretty unreasonable. But I think for a 0.3x-0.5x range, the latter could end up reasonable while the former gives the user the impression that something is broken (similar to the brokenness you can encounter when entering "desktop web mode" on a phone, or when pages have misconfigured viewport metas).

**Comment by jeremyroman** [comment]
My reasons for preferring the former are:

1. don't need to relayout/repaint on activation (or worse, every frame of an intervening animation)
2. doesn't allow the host document to control the inner viewport, which could be used to communicate to it

**Comment by domenic** [comment]
Yep, I totally agree, I just think that if we are envisioning scenarios like some of those in the mocks in https://github.com/WICG/portals/blob/master/key-scenarios.md , pages will need a styling hook more so than they did with the latter.

**Comment by bokand** [comment]
IMHO, to achieve the use cases above, the onus should be on the destination page to provide the preview. I've outlined how that might look in #241. tl;dr The embeddee may not want anyone to control the preview, changes to the portaled page mean the embedder's customizations may break, and the embedder probably can't provide a great preview anyway. We should define a sensible default and give the destination page a hook to define a good preview.

If a page is especially badly suited, like in @domenic's example above), the embedder could always just slap a GitHub logo and some text over top of the portal and activate it on click.

WDYT? I think the styling hooks aren't necessarily an MVP but I'd like to start implementing the change to use a replaced-like rendering model and fixing the layout size to the outer page's viewport. Does anyone have major objections to doing that?

**Comment by domenic** [comment]
I agree with your conclusions and proposed path forward. Now to go look at #241.
